zcash_client_sqlite/src/wallet/sapling.rs

```diff
@@ -136,37 +136,50 @@ fn to_spendable_note<P: consensus::Parameters>(
             SqliteClientError::CorruptedData("Note commitment tree position invalid.".to_string())
         })?);
 
-    let ufvk_str: String = row.get(7)?;
-    let ufvk = UnifiedFullViewingKey::decode(params, &ufvk_str)
-        .map_err(SqliteClientError::CorruptedData)?;
-
-    let scope_code: i64 = row.get(8)?;
-    let spending_key_scope = parse_scope(scope_code).ok_or_else(|| {
-        SqliteClientError::CorruptedData(format!("Invalid key scope code {}", scope_code))
-    })?;
-
-    let recipient = match spending_key_scope {
-        Scope::Internal => ufvk
-            .sapling()
-            .and_then(|dfvk| dfvk.diversified_change_address(diversifier)),
-        Scope::External => ufvk
-            .sapling()
-            .and_then(|dfvk| dfvk.diversified_address(diversifier)),
-    }
-    .ok_or_else(|| SqliteClientError::CorruptedData("Diversifier invalid.".to_owned()))?;
-
-    Ok(ReceivedNote::from_parts(
-        note_id,
-        txid,
-        output_index,
-        Note::Sapling(sapling::Note::from_parts(
-            recipient,
-            sapling::value::NoteValue::from_raw(note_value),
-            rseed,
-        )),
-        spending_key_scope,
-        note_commitment_tree_position,
-    ))
+    let ufvk_str: Option<String> = row.get(7)?;
+    let scope_code: Option<i64> = row.get(8)?;
+
+    // If we don't have information about the recipient key scope or the ufvk we can't determine
+    // which spending key to use. This may be because the received note was associated with an
+    // imported viewing key, so we treat such notes as not spendable. Although this method is
+    // presently only called using the results of queries where both the ufvk and
+    // recipient_key_scope columns are checked to be non-null, this is method is written
+    // defensively to account for the fact that both of these are nullable columns in case it
+    // is used elsewhere in the future.
+    ufvk_str
+        .zip(scope_code)
+        .map(|(ufvk_str, scope_code)| {
```

Comment:
We can always determine the scope when we have a UFVK _and are scanning with it_, so it could be considered an `SqliteClientError::CorruptedData` if one is null and the other is non-null, but it could also be either a programming error on the part of the caller, or a database that had a UIVK imported and then later a corresponding UFVK imported (which is not a use case I am advocating we support). So from a defensive implementation perspective I think this code is fine (we just ignore these notes instead of erroring).